blockchain/upgrade.go

```diff
@@ -2742,173 +2742,192 @@ func migrateSpendJournalVersion2To3(ctx context.Context, b *BlockChain) error {
 					//     script version      VLQ      variable
 					//     compressed script   []byte   variable
+					//   stakeExtra            []byte   variable
 					//
-					//   OPTIONAL
-					//     ticketMinOuts      []byte         variable
+					// The serialized flags code format is:
+					//   bit  0   - containing transaction is a coinbase
+					//   bit  1   - containing transaction has an expiry
+					//   bits 2-4 - transaction type
+					//   bit  5   - unused
+					//   bit  6   - is fully spent
+					//   bit  7   - unused
 					//
-					// The serialized flags format is:
-					//   bit  0     - containing transaction is a coinbase
-					//   bit  1     - containing transaction has an expiry
-					//   bits 2-5   - transaction type
-					//   bits 6-7   - unused
+					// The serialized header code format is:
+					//   bit 0 - output zero is unspent
+					//   bit 1 - output one is unspent
+					//   bits 2-x - number of bytes in unspentness bitmap.  When both bits
+					//     1 and 2 are unset, it encodes N-1 since there must be at least
+					//     one unspent output.
 					//
-					// The ticket min outs field contains minimally encoded outputs for
-					// all outputs of a ticket transaction. It is only encoded for ticket
-					// submission outputs.
-					utxoSerialized := cursor.Value()
+					// The stake extra field contains minimally encoded outputs for all
+					// consensus-related outputs in the stake transaction. It is only
+					// encoded for tickets.
 
-					// Deserialize the block height.  Ignore it since we don't need it.
+					// Deserialize the version.  Ignore it since we don't need it.
 					_, bytesRead = deserializeVLQ(utxoSerialized)
 					utxoOffset := bytesRead
+					if utxoOffset >= len(utxoSerialized) {
+						return false, errDeserialize("unexpected end of data after version")
+					}
+
+					// Deserialize the block height.  Ignore it since we don't need it.
+					_, bytesRead = deserializeVLQ(utxoSerialized[utxoOffset:])
+					utxoOffset += bytesRead
 					if utxoOffset >= len(utxoSerialized) {
 						return false, errDeserialize("unexpected end of data after height")
 					}
 
-					// Deserialize the block index.  Ignore it since we don't need it.
+					// Deserialize the block index. Ignore it since we don't need it.
 					_, bytesRead = deserializeVLQ(utxoSerialized[utxoOffset:])
 					utxoOffset += bytesRead
 					if utxoOffset >= len(utxoSerialized) {
 						return false, errDeserialize("unexpected end of data after index")
 					}
 
 					// Deserialize the flags.
-					v3UtxoFlags, bytesRead := deserializeVLQ(utxoSerialized[utxoOffset:])
+					v2UtxoFlags, bytesRead := deserializeVLQ(utxoSerialized[utxoOffset:])
 					utxoOffset += bytesRead
 					if utxoOffset >= len(utxoSerialized) {
 						return false, errDeserialize("unexpected end of data after flags")
 					}
 
 					// Set the flags on the stxo.
-					stxo.flags = uint8(v3UtxoFlags)
+					stxo.flags = uint8(v2UtxoFlags)
```

Comment:
I believe this should be decoding and re-encoding the flags here because the v3 stxo flag format is not the same as the v2 utxo flag format.  It is probably skating by because the only observable difference seems to be the utxo entry spent bit (bit 6) which is unused in the v3 stxo flag format, but I think it would be safer to actually re-encode the flags properly here to avoid leaving that stray bit set when it really shouldn't be.